### PR TITLE
Feature: Delete Todo

### PR DESCRIPTION
**Pull request by ich-wtag**
Overview

[Task-2: [Delete Todo](https://trello.com/c/IKF44wvt/2-task-2-delete-todo)
This PR introduces the functionality of delete a todo by clicking the delete button beside input box.

Types of Changes

 

- [x] Todo can be deleted
- [x] A list is populated with the updated Todo

Review
Please have a close look on this PR and kindly put your valuable comments.

[delete-todo_vanilla-todo.webm](https://github.com/ich-wtag/Todo-App_Vanilla_Js/assets/153156568/1e9f36a1-0d62-4ce9-a097-df44944a7c9d)

### CODE REVIEW

File: script.js [hunk 1]
```diff
@@ -23,17 +23,32 @@ const addTodoHandler = () => {
     renderTodos(todos);
 };
 
+const deleteTodoHandler = (todoId) => {
+    todos = [...todos].filter((todo) => todo.id !== todoId);
```
Comment (hhu-wtag) [inline script.js]: No need to spread as spreading is an `O(n)` operation. Also there is no issues or reference.
```suggestion
    todos = todos.filter((todo) => todo.id !== todoId);
```

File: script.js [hunk 2]
```diff
@@ -23,17 +23,32 @@ const addTodoHandler = () => {
...
+
+    renderTodos(todos);
+};
+
 const createTodoElement = (todo) => {
     const $todo = document.createElement("li");
+    const $deleteButton = document.createElement("button");
+    $deleteButton.innerText = "Delete";
+
+    eventListenerHandler($deleteButton, "click", () =>
```
Comment (hhu-wtag) [inline script.js]: what does this function do? Is this is factory function? If so let's add a verb to indicate that it does some work.
e.g: **createEventListener**
Also I don't think that we need to separate the logic into another function.
```js
$deleteButton.addEventListener("click", () => deleteTodoHandler(todo.id))
```

works fine.

File: script.js [hunk 3]
```diff
@@ -23,17 +23,32 @@ const addTodoHandler = () => {
...
+    return $todo;
 };
 
 const renderTodos = (todos) => {
     $todoList.innerHTML = "";
-
-    $todoList.innerHTML = todos.map((todo) => createTodoElement(todo)).join("");
+    todos.forEach((todo) => {
+        $todoList.appendChild(createTodoElement(todo));
+    });
```
Comment (hhu-wtag) [inline script.js:45]: Why are we opting back to previous suggestion?

**Review by hhu-wtag — CHANGES_REQUESTED** [review]

File: script.js (hunk 3 above)
Comment (ich-wtag) [reply script.js:45]: As our createElement create new element. If we use the map method to render the todos it just return us Object html Li element. so I had to make it  to string using outerHTml. But  when I dynamically added button and eventListeners, it also  converted into string. So finally I can't find the events. For this we have to use forEach method. As each time we will append a new child by this. Also we don't have to use outerHTML too.

File: script.js (hunk 2 above)
Comment (ich-wtag) [reply script.js]: Bhai as we have so many eventListeners, For that I created a factory function for the eventListener in utility.js file. Okay I'll remove the function.

**Review by hhu-wtag — APPROVED** [review]

**Pull request merged**